unittests: Fixes unit tests for Windows (part 4)

claudiubelu opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind failing-test

/sig windows

#### What this PR does / why we need it:

Currently, there are some unit tests that are failing on Windows due to various reasons:

- paths not properly joined (``filepath.Join`` should be used).
- files not closed, which means that they cannot be removed / renamed.
- ``time.Now()`` is not as precise on Windows, which means that 2 consecutive calls may return the same timestamp.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Related: #51540

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

claudiubelu committed:
unittests: Fixes unit tests for Windows (part 4)

Currently, there are some unit tests that are failing on Windows due to
various reasons:

- paths not properly joined (filepath.Join should be used).
- files not closed, which means that they cannot be removed / renamed.
- time.Now() is not as precise on Windows, which means that 2
  consecutive calls may return the same timestamp.

endocrimes commented:
/assign

lianghao208 commented on pkg/kubelet/kubeletconfig/configfiles/configfiles.go:
```diff
@@ -83,8 +84,18 @@ func resolveRelativePaths(paths []*string, root string) {
+
+	// on Windows, filepath.IsAbs will not return True for paths prefixed with a slash, even
+	// though they can be used as absolute paths.
+	return goruntime.GOOS == "windows" && len(path) > 0 && (path[0] == '\\' || path[0] == '/')
```
Is this approach necessary?  [src/path/filepath/path_windows.go#L40](https://github.com/golang/go/blob/master/src/path/filepath/path_windows.go#L49)

Can you elaborate in which situation windows path with "/" or "" considered as absolute paths?

claudiubelu replied:
Sure. At the link you've given, it checks if it starts with ``\\`` or ``//`` exactly, but not for simply ``\`` or ``/``. ``filepath.IsAbs`` does not consider those as absolute paths, even though Windows does: https://docs.microsoft.com/en-us/dotnet/standard/io/file-path-formats

SergeyKanzhelev commented on pkg/kubelet/kuberuntime/helpers.go:
```diff
@@ -238,7 +238,7 @@ func annotationProfile(profile, profileRootPath string) string {
 	if strings.HasPrefix(profile, v1.SeccompLocalhostProfileNamePrefix) {
 		name := strings.TrimPrefix(profile, v1.SeccompLocalhostProfileNamePrefix)
 		fname := filepath.Join(profileRootPath, filepath.FromSlash(name))
```
I wonder what `profileRootPath` will be on windows. Can it have `c:\` prefix?

claudiubelu replied:
Hm, we don't support seccomps on Windows, so we should probably skip those tests anyways.

claudiubelu commented:
/retest

SergeyKanzhelev commented:
/priority important-soon

I'd let sig windows to triage it

jsturtevant commented:
/triage accepted

claudiubelu commented:
/test pull-kubernetes-e2e-kind-ipv6

jsturtevant commented:
for context, we recently added unit tests CI for Windows and this fixes a subset of the failing tests: https://testgrid.k8s.io/sig-windows-signal#windows-unit-master

claudiubelu commented:
/sig testing

endocrimes commented on pkg/volume/flexvolume/probe.go:
```diff
@@ -173,7 +173,7 @@ func (prober *flexVolumeProber) handleWatchEvent(event fsnotify.Event) error {
 				return err
 			}
-			if err := prober.addWatchRecursive(pluginDirAbs); err != nil {
+			if err := prober.addWatchRecursive(prober.pluginDir); err != nil {
```
not sure why we're changing this away from being the absolute path?

claudiubelu replied:
Reverted.

marosset commented:
@claudiubelu can we make a new PR w/ just the log rotation fixes and try and get that merged into v1.25?

claudiubelu commented:
> @claudiubelu can we make a new PR w/ just the log rotation fixes and try and get that merged into v1.25?

Done: https://github.com/kubernetes/kubernetes/pull/111549

claudiubelu commented:
/milestone v1.26

claudiubelu commented:
/assign @dchen1107 @jingxu97 @kubernetes/sig-scheduling @kubernetes/sig-storage-pr-reviews

SergeyKanzhelev commented on pkg/kubelet/kubeletconfig/configfiles/util_windows.go:
```diff
@@ -0,0 +1,30 @@
+func isAbs(path string) bool {
+	// on Windows, filepath.IsAbs will not return True for paths prefixed with a slash, even
+	// though they can be used as absolute paths (https://docs.microsoft.com/en-us/dotnet/standard/io/file-path-formats).
+	return filepath.IsAbs(path) || (len(path) > 0 && (path[0] == '\\' || path[0] == '/'))
```
is it an error in `filepath` package logic? Is there a github issue where this might have been discussed? Looks like a common, non k8s specific problem.

claudiubelu replied:
From the ``filepath`` package logic itself, it seems that it won't return ``True`` for any slash-prefixed paths (e.g.: ``\foo``), since ``volumeNameLen`` would return 0 (doesn't start with a drive letter, and it's not an UNC path), which means that ``filepath.IsAbs`` returns ``False``. Though technically, according to https://docs.microsoft.com/en-us/dotnet/standard/io/file-path-formats , it should be considered as absolute. According to the same documentation, ``\foo`` would be normalized to ``C:\foo`` (``If the path starts with a single component separator, the drive from the current directory is applied. For example, if the file path is \utilities and the current directory is C:\temp\, normalization produces C:\utilities.``). This information can also be seen here: https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#fully-qualified-vs-relative-paths (``Fully Qualified vs. Relative Paths - A file name is relative to the current directory if it does not begin with one of the following: ... A single backslash, for example, "\directory" or "\file.txt". This is also referred to as an absolute path.``)

It might also be worth noting that ``filepath.Abs("/foo")`` (which is a ``GetFullPathName`` syscall) will result in normalized ``C:\foo`` path, which applies the rule above.

SergeyKanzhelev replied:
my question is - if this is not k8s specific issue, can we file an issue for the `filepath` package itself? Or is there any k8s-specific here?

SergeyKanzhelev commented:
this PR changes code logic as well. I'd suggest to retitle this PR to clarify it.

Also PR has unanswered comment from @endocrimes, can you please address comments?

claudiubelu commented:
> this PR changes code logic as well. I'd suggest to retitle this PR to clarify it.
> 
> Also PR has unanswered comment from @endocrimes, can you please address comments?

I've addressed those comments.

I've also split a few changes out of this PR, specifically to the found issues:

Windows: Consider slash-prefixed paths as absolute: https://github.com/kubernetes/kubernetes/pull/112566
Fixes getNestedMountpoints grouping: https://github.com/kubernetes/kubernetes/pull/112571

SergeyKanzhelev commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -1022,7 +1022,9 @@ func (f *frameworkImpl) runPreBindPlugin(ctx context.Context, pl framework.PreBi
+	// resulting in 0 time delta / latency.
+	startTime := time.Now().Add(-time.Microsecond)
 	defer func() {
 		metrics.FrameworkExtensionPointDuration.WithLabelValues(bind, status.Code().String(), f.profileName).Observe(metrics.SinceInSeconds(startTime))
```
is zero duration ever desirable in `SinceInSeconds`? Can we just implement a generic logic there to avoid the need to patch every place it may become an issue on Windows?

claudiubelu replied:
Done

SergeyKanzhelev commented on staging/src/k8s.io/mount-utils/mount_windows.go:
```diff
@@ -300,13 +309,13 @@ func (mounter *SafeFormatAndMount) formatAndMountSensitive(source string, target
+		return NewMountError(UnknownMountError, err.Error())
 	}
-	klog.V(2).Infof("formatAndMount disk(%s) fstype(%s) on(%s) with output(%s) successfully", driverPath, fstype, target, string(output))
+
```
is it intended change of logging level from 2 to 4?

claudiubelu replied:
Excluded change. Will revisit this in a later PR.

SergeyKanzhelev reviewed (approved):
/lgtm

claudiubelu commented:
/test pull-kubernetes-e2e-capz-windows-containerd
/assign @damemi @saad-ali

jsafrane commented:
/approve
storage parts

damemi reviewed (commented):
lgtm from scheduler
cc @ahg-g for visibility

marosset commented:
/assign @alculquicondor 
for pkg/scheduler review

alculquicondor reviewed (commented):
This is very hacky.
Is there an open issue in golang or wherever it's appropriate?

We should at least put a TODO to remove this hack once the underlying problem is solved.

claudiubelu commented:
> This is very hacky. Is there an open issue in golang or wherever it's appropriate?
> 
> We should at least put a TODO to remove this hack once the underlying problem is solved.

You mean the slash-prefixed paths to be considered absolute? I've moved that part to a separate PR, so this PR doesn't get blocked on that issue. Will add the TODO there.

claudiubelu commented:
> I mean `time.Since`

I've added a reference to https://github.com/golang/go/issues/8687 in ``SinceInSeconds`` function. The Windows clock resolution is a known issue, and it's not an easy thing to change, as it can have quite a few implications on how applications behave, power consumption, etc., especially since it's a global option.

claudiubelu commented:
/test pull-kubernetes-e2e-capz-windows-containerd

alculquicondor commented on pkg/scheduler/metrics/util_windows.go:
```diff
@@ -0,0 +1,32 @@
+func SinceInSeconds(start time.Time) float64 {
+	// On Windows, time.Now() is not as precise, 2 consecutive calls may return the same timestamp,
+	// resulting in 0 time delta / latency.
+	// See: https://github.com/golang/go/issues/8687
```
The issue is closed.

Why is 0 delta a problem?

claudiubelu replied:
Test assertions are expecting the latency to be non-zero: https://github.com/kubernetes/kubernetes/blob/70617042976dc168208a41b8a10caa61f9748617/pkg/scheduler/framework/runtime/framework_test.go#L2815

The metrics latencies are registered whenever certain operations complete (e.g.: ``RunPreFilterPlugins``), and ``SinceIsSeconds`` is used to calculate these deltas (e.g.: https://github.com/kubernetes/kubernetes/blob/70617042976dc168208a41b8a10caa61f9748617/pkg/scheduler/framework/runtime/framework.go#L607 ). But since ``time.Now()`` can return the same timestamp on Windows, the latency becomes 0.

These are just some measurements aggregated into histograms. Alternatively, we can skip the check ``if value <= 0`` mentioned above for Windows.

alculquicondor replied:
if that's easy to achieve, that would make more sense.

claudiubelu replied:
@alculquicondor , done. In the current version of the PR we're skipping the ``if value <= 0`` check.

claudiubelu commented:
/milestone v1.27

claudiubelu commented:
/retest

claudiubelu commented:
@alculquicondor , can you take another look?

alculquicondor reviewed (commented):
/approve

claudiubelu commented:
> /approve

Could you also lgtm pls?
